File: old/services/ssh.py
"""
SSH Service - manages SSH connections and command execution
"""
import logging
import sys
import time
from pathlib import Path
from typing import Optional
try:
    import paramiko
    HAS_PARAMIKO = True
except ImportError:
    HAS_PARAMIKO = False
from libs.config import SSHConfig
logger = logging.getLogger(__name__)

class SSHService:
    """Service that manages SSH connections and command execution"""
    def __init__(self, host: str, ssh_config: SSHConfig):
        """
        Initialize SSH service
        Args:
            host: SSH host (format: user@host or just host)
            ssh_config: SSH configuration
        """
        self.host = host
        self.ssh_config = ssh_config
        self._client: Optional[paramiko.SSHClient] = None
        self._connected = False
        # Parse host (format: user@host or just host)
        if "@" in host:
            self.username, self.hostname = host.split("@", 1)
        else:
            self.username = ssh_config.default_username
            self.hostname = host

    def connect(self) -> bool:
        """
        Establish SSH connection
        Returns:
            True if connection successful, False otherwise
        """
        if self._connected and self._client:
            # Check if connection is still alive
            try:
                transport = self._client.get_transport()
                if transport and transport.is_active():
                    return True
            except Exception:
                # Connection is dead, need to reconnect
                self._client = None
                self._connected = False
        if not HAS_PARAMIKO:
            logger.error("paramiko not available, cannot establish SSH connection")
            return False
        try:
            self._client = paramiko.SSHClient()
            self._client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
            # Find and load private key that matches the public key we set up
            # We set up id_rsa.pub or id_ed25519.pub, so try to load the corresponding private key
            key_file = None
            key_paths = [
                Path.home() / ".ssh" / "id_rsa",
                Path.home() / ".ssh" / "id_ed25519",
            ]
            for key_path in key_paths:
                if key_path.exists():
                    try:
                        key_file = str(key_path)
                        logger.debug("Found private key: %s", key_file)
                        break
                    except Exception:
                        continue
            connect_kwargs = {
                "hostname": self.hostname,
                "username": self.username,
                "timeout": self.ssh_config.connect_timeout,
                "look_for_keys": self.ssh_config.look_for_keys,
                "allow_agent": self.ssh_config.allow_agent,
            }
            # Explicitly load private key if found
            if key_file:
                try:
                    # Try to load the key
                    pkey = None
                    for key_class in [paramiko.RSAKey, paramiko.Ed25519Key]:
                        try:
                            pkey = key_class.from_private_key_file(key_file)
                            logger.debug("Loaded private key from %s", key_file)
                            break
                        except Exception:
                            continue
                    if pkey:
                        connect_kwargs["pkey"] = pkey
                except Exception as key_exc:
                    logger.warning("Failed to load private key %s: %s", key_file, key_exc)
            self._client.connect(**connect_kwargs)
            self._connected = True
            logger.info("SSH connection established to %s@%s", self.username, self.hostname)
            return True
        except paramiko.AuthenticationException as exc:
            logger.error("SSH authentication failed to %s: %s", self.host, exc)
            logger.error("Tried key file: %s", key_file if "key_file" in locals() else "None")
            self._client = None
            self._connected = False
            return False
        except paramiko.SSHException as exc:
            logger.error("SSH connection error to %s: %s", self.host, exc)
            self._client = None
            self._connected = False
            return False
        except Exception as exc:
            logger.error("Failed to establish SSH connection to %s: %s", self.host, exc)
            logger.error("Exception type: %s", type(exc).__name__)
            self._client = None
            self._connected = False
            return False

    def disconnect(self):
        """Close SSH connection"""
        if self._client:
            try:
                self._client.close()
            except Exception:
                pass
            finally:
                self._client = None
                self._connected = False
            logger.debug("SSH connection closed to %s", self.host)

    def is_connected(self) -> bool:
        """Check if SSH connection is active"""
        if not self._connected or not self._client:
            return False
        try:
            transport = self._client.get_transport()
            return transport is not None and transport.is_active()
        except Exception:
            return False

    def execute(self, command: str, timeout: Optional[int] = None, sudo: bool = False):
        """
        Execute command via SSH connection (always shows output interactively and captures it)
        Args:
            command: Command to execute
            timeout: Command timeout in seconds
            sudo: Whether to run command with sudo
        Returns:
            Tuple of (output, exit_code). output is always captured
        """
        if not self.is_connected():
            if not self.connect():
                logger.error("Cannot execute command: SSH connection not available")
                return None, None
        if sudo:
            # For multi-line scripts, use base64 encoding to avoid quoting issues
            if "\n" in command:
                import base64
                encoded = base64.b64encode(command.encode("utf-8")).decode("utf-8")
                command = f"sudo -n bash -c 'echo {encoded} | base64 -d | bash'"
            else:
                # For single-line commands, wrap in bash -c with proper quoting
                import shlex
                escaped_command = shlex.quote(command)
                command = f"sudo -n bash -c {escaped_command}"
        logger.debug("Running: %s", command)
        exec_timeout = timeout if timeout else self.ssh_config.default_exec_timeout
        try:
            _, stdout, stderr = self._client.exec_command(command, timeout=exec_timeout, get_pty=True)
            channel = stdout.channel
            # Read output
            output_lines = []
            error_lines = []
            last_output_time = time.time()
            # Set channel to non-blocking
            channel.setblocking(0)
            while True:
                current_time = time.time()
                received_output = False
                # Check if channel is closed and exit status is ready
                if channel.exit_status_ready():
                    break
                # Read available data from stdout
                if channel.recv_ready():
                    try:
                        data = stdout.read(self.ssh_config.read_buffer_size).decode("utf-8", errors="replace")
                        if data:
                            received_output = True
                            last_output_time = current_time
                            output_lines.append(data)
                            if self.ssh_config.verbose:
                                sys.stdout.write(data)
                                sys.stdout.flush()
                                logger.info(data.rstrip())
                    except (IOError, OSError, UnicodeDecodeError):
                        pass
                # Read available data from stderr
                if channel.recv_stderr_ready():
                    try:
                        data = stderr.read(self.ssh_config.read_buffer_size).decode("utf-8", errors="replace")
                        if data:
                            received_output = True
                            last_output_time = current_time
                            error_lines.append(data)
                            if self.ssh_config.verbose:
                                sys.stderr.write(data)
                                sys.stderr.flush()
                                logger.error(data.rstrip())
                    except (IOError, OSError, UnicodeDecodeError):
                        pass
                # Check for timeout
                if not received_output:
                    time_since_last_output = current_time - last_output_time
                    if time_since_last_output > exec_timeout:
                        logger.error("SSH command timeout after %ss of no output - COMMAND FAILED", exec_timeout)
                        channel.close()
                        return None, None
                time.sleep(self.ssh_config.poll_interval)
            # Get remaining output
            try:
                remaining = stdout.read().decode("utf-8", errors="replace")
                if remaining:
                    output_lines.append(remaining)
                    if self.ssh_config.verbose:
                        sys.stdout.write(remaining)
                        sys.stdout.flush()
                        logger.info(remaining.rstrip())
                remaining_err = stderr.read().decode("utf-8", errors="replace")
                if remaining_err:
                    error_lines.append(remaining_err)
                    if self.ssh_config.verbose:
                        sys.stderr.write(remaining_err)
                        sys.stderr.flush()
                        logger.error(remaining_err.rstrip())
            except (IOError, OSError, UnicodeDecodeError):
                pass
            # Get exit status
            exit_code = channel.recv_exit_status()
            output = "".join(output_lines).strip()
            error_output = "".join(error_lines).strip()
            # Combine stdout and stderr
            combined = output
            if error_output:
                combined = f"{output}\n{error_output}" if output else error_output
            return combined, exit_code
        except paramiko.SSHException as exc:
            logger.error("SSH command execution failed: %s", exc)
            return None, None
        except Exception as exc:
            logger.error("Unexpected error during SSH command execution: %s", exc)
            return None, None

    def __enter__(self):
        """Context manager entry"""
        self.connect()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        """Context manager exit"""
        self.disconnect()

File: src/services/ssh.go
package services

import (
	"encoding/base64"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"golang.org/x/crypto/ssh"

	"enva/libs"
)

// SSHService manages SSH connections and command execution
type SSHService struct {
	host      string
	hostname  string
	username  string
	sshConfig *libs.SSHConfig
	client    *ssh.Client
	connected bool
}

// NewSSHService creates a new SSH service
func NewSSHService(host string, sshConfig *libs.SSHConfig) *SSHService {
	service := &SSHService{
		host:      host,
		sshConfig: sshConfig,
		connected: false,
	}

	// Parse host (format: user@host or just host)
	if idx := strings.Index(host, "@"); idx >= 0 {
		service.username = host[:idx]
		service.hostname = host[idx+1:]
	} else {
		service.username = sshConfig.DefaultUsername
		service.hostname = host
	}

	return service
}

// Connect establishes SSH connection
func (s *SSHService) Connect() bool {
	if s.connected && s.client != nil {
		// Check if connection is still alive
		_, _, err := s.client.SendRequest("keepalive", false, nil)
		if err == nil {
			return true
		}
		// Connection is dead, need to reconnect
		s.client = nil
		s.connected = false
	}

	// Find and load private key
	keyFile := s.findPrivateKey()
	if keyFile == "" {
		libs.GetLogger("ssh").Printf("No private key found")
		return false
	}

	// Load private key
	signer, err := s.loadPrivateKey(keyFile)
	if err != nil {
		libs.GetLogger("ssh").Printf("Failed to load private key %s: %v", keyFile, err)
		return false
	}

	// Create SSH config
	config := &ssh.ClientConfig{
		User:            s.username,
		Auth:            []ssh.AuthMethod{ssh.PublicKeys(signer)},
		HostKeyCallback: ssh.InsecureIgnoreHostKey(), // In production, use knownhosts
		Timeout:         time.Duration(s.sshConfig.ConnectTimeout) * time.Second,
	}

	// Connect
	addr := fmt.Sprintf("%s:22", s.hostname)
	client, err := ssh.Dial("tcp", addr, config)
	if err != nil {
		libs.GetLogger("ssh").Printf("Failed to establish SSH connection to %s: %v", s.host, err)
		return false
	}

	s.client = client
	s.connected = true
	libs.GetLogger("ssh").Printf("SSH connection established to %s@%s", s.username, s.hostname)
	return true
}

// Disconnect closes SSH connection
func (s *SSHService) Disconnect() {
	if s.client != nil {
		s.client.Close()
		s.client = nil
		s.connected = false
		libs.GetLogger("ssh").Printf("SSH connection closed to %s", s.host)
	}
}

// IsConnected checks if SSH connection is active
func (s *SSHService) IsConnected() bool {
	if !s.connected || s.client == nil {
		return false
	}
	_, _, err := s.client.SendRequest("keepalive", false, nil)
	return err == nil
}

// Execute executes command via SSH connection
func (s *SSHService) Execute(command string, timeout *int, sudo ...bool) (string, *int) {
	if !s.IsConnected() {
		if !s.Connect() {
			libs.GetLogger("ssh").Printf("Cannot execute command: SSH connection not available")
			return "", nil
		}
	}

	execTimeout := s.sshConfig.DefaultExecTimeout
	if timeout != nil {
		execTimeout = *timeout
	}

	useSudo := len(sudo) > 0 && sudo[0]
	if useSudo {
		// For multi-line scripts or commands with single quotes, use base64 encoding to avoid quoting issues (matching Python behavior)
		if strings.Contains(command, "\n") || strings.Contains(command, "'") {
			encoded := base64.StdEncoding.EncodeToString([]byte(command))
			command = fmt.Sprintf("sudo -n bash -c 'echo %s | base64 -d | bash'", encoded)
		} else {
			// For single-line commands without single quotes, wrap in bash -c with proper quoting (matching Python shlex.quote behavior)
			command = fmt.Sprintf("sudo -n bash -c %s", quoteCommand(command))
		}
	}

	libs.GetLogger("ssh").Debug("Running: %s", command)

	// Create session
	session, err := s.client.NewSession()
	if err != nil {
		libs.GetLogger("ssh").Printf("Failed to create SSH session: %v", err)
		return "", nil
	}
	defer session.Close()

	// Set up terminal
	modes := ssh.TerminalModes{
		ssh.ECHO:          0,
		ssh.TTY_OP_ISPEED: 14400,
		ssh.TTY_OP_OSPEED: 14400,
	}
	if err := session.RequestPty("xterm", 80, 40, modes); err != nil {
		libs.GetLogger("ssh").Printf("Failed to request PTY: %v", err)
	}

	// Set up stdout/stderr
	var stdout, stderr strings.Builder
	session.Stdout = &stdout
	session.Stderr = &stderr

	// Execute command with timeout
	done := make(chan error, 1)
	go func() {
		done <- session.Run(command)
	}()

	select {
	case err := <-done:
		exitCode := 0
		if err != nil {
			if exitErr, ok := err.(*ssh.ExitError); ok {
				exitCode = exitErr.ExitStatus()
			} else {
				libs.GetLogger("ssh").Printf("Command execution error: %v", err)
				return "", nil
			}
		}
		output := stdout.String()
		errorOutput := stderr.String()
		combined := output
		if errorOutput != "" {
			if output != "" {
				combined = output + "\n" + errorOutput
			} else {
				combined = errorOutput
			}
		}
		// Show output if verbose
		if s.sshConfig.Verbose {
			fmt.Print(combined)
		}
		return strings.TrimSpace(combined), &exitCode
	case <-time.After(time.Duration(execTimeout) * time.Second):
		session.Close()
		libs.GetLogger("ssh").Printf("SSH command timeout after %ds - COMMAND FAILED", execTimeout)
		return "", nil
	}
}

// findPrivateKey finds private key file
func (s *SSHService) findPrivateKey() string {
	homeDir, err := os.UserHomeDir()
	if err != nil {
		return ""
	}

	keyPaths := []string{
		filepath.Join(homeDir, ".ssh", "id_rsa"),
		filepath.Join(homeDir, ".ssh", "id_ed25519"),
	}

	for _, keyPath := range keyPaths {
		if _, err := os.Stat(keyPath); err == nil {
			return keyPath
		}
	}
	return ""
}

// loadPrivateKey loads private key from file
func (s *SSHService) loadPrivateKey(keyFile string) (ssh.Signer, error) {
	key, err := os.ReadFile(keyFile)
	if err != nil {
		return nil, err
	}

	// Try RSA key first
	signer, err := ssh.ParsePrivateKey(key)
	if err == nil {
		return signer, nil
	}

	// Try with passphrase (empty for now)
	signer, err = ssh.ParsePrivateKeyWithPassphrase(key, []byte(""))
	if err == nil {
		return signer, nil
	}

	return nil, fmt.Errorf("failed to parse private key: %v", err)
}

// quoteCommand quotes a command for use in bash -c (matching Python shlex.quote)
func quoteCommand(cmd string) string {
	// Simple quoting - wrap in single quotes and escape single quotes
	if strings.Contains(cmd, "'") || strings.Contains(cmd, " ") || strings.Contains(cmd, "$") {
		escaped := strings.ReplaceAll(cmd, "'", "'\"'\"'")
		return fmt.Sprintf("'%s'", escaped)
	}
	return fmt.Sprintf("'%s'", cmd)
}
